### DISCUSSION TITLE
Fix npm security vulnerabilities

### DISCUSSION THREAD

**Discussion by lawdank** [uwrit/leaf#559] [Q&A]
We are using **_Leaf v3.11.1_** & the related _**npm version 6.14.17**_
npm related Security vulnerabilities have been flagged for us. 

Using command - _npm audit fix_ 
**fixed 6 of 59 vulnerabilities in 1772 scanned packages
1 vulnerability required manual review and could not be updated
1 package update for 52 vulnerabilities involved breaking changes**

On checking further [using command - _npm audit_]:
   === npm audit security report ===

    Run npm install react-scripts@5.0.1 to resolve 52 vulnerabilities
    SEMVER WARNING: Recommended action is a potentially breaking change

Since, this could be a potentially breaking change. Do you have a recommended version of npm that we can safely upgrade to or upgrading single package _react-script_ is safe or something else?

**Comment by artgoldberg** [comment]
Hi @ndobb and @mh2727 . lawdank is my colleague Sheersha reverse(Kandwal).

Restating her question, how far can we advance the `npm` version without breaking the Leaf GUI?

Thanks
Arthur

**Comment by ndobb** [comment]
@lawdank and @artgoldberg  I will try to test things out tomorrow and let you know. Upgrading JavaScript dependencies with `npm` tends to be a never-ending game a whack-a-mole.

**Reply by artgoldberg** [reply]
@ndobb Thanks for investigating this! Whack-a-mole not a fun game.

**Reply by ndobb** [reply]
@lawdank and @artgoldberg okay after an afternoon of working through various breaking changes and so on, I've been able to eliminate all security warnings on the client. I'm working on this through the [active development branch](https://github.com/uwrit/leaf/tree/column-rename) on I'm on right now, and you can see the various changes (if you're interested) in [this commit](https://github.com/uwrit/leaf/commit/7c4b96b6bbb756fedf135914d71f8fd35baf3300).

So to reiterate, yes, unfortunately these are breaking changes and require editing Leaf's source code. If you do a `npm audit fix -f`, the app will not compile afterward.

I'm working to finalize the changes in the current dev branch, but until then the dependency changes unfortunately won't be available. As an aside, nearly all of the issues were due to internal dependencies within the `react-scripts` package. Upgrading that resolved all but a single issue (and also caused the breaking changes). [`react-scripts was not previously`](https://github.com/uwrit/leaf/blob/master/src/ui-client/package.json#L39) a so-called [`devDependency` in the web app configuration](https://github.com/uwrit/leaf/blob/master/src/ui-client/package.json#L56), but should have been, as it's not actually used at all after production compilation. So in other words, the security issues listed shouldn't affect a production environment to the best of my knowledge.

One more aside that I learned as well: if you want to only check issues that would affect a production environment, you can use `npm audit --production`. Doing so in my current branch finds no issues:

![image](https://user-images.githubusercontent.com/10424962/189243341-9c328658-57ec-4702-b2c8-5f9e5b45b75d.png)

Doing so without the `--production` flag finds a single issue:

![image](https://user-images.githubusercontent.com/10424962/189243767-4e9b7c64-0ebd-4129-a6b7-e2b799ccab85.png)

But [this too appears due to `react-scripts`](https://github.com/facebook/create-react-app/issues/11647), which is as I mentioned not needed in production.

**Reply by artgoldberg** [reply]
Hello @ndobb & @lawdank 
@ndobb, we greatly appreciate your attention to our concerns about `npm` vulnerabilities. Nice work on resolving them and thank you for providing a detailed explanation of the relationships between `npn` and `react-scripts`, and their roles in dev and production.
I expect that our security folks will be reassured that "the security issues listed shouldn't affect a production environment (to the best of Nic's knowledge)."
Arthur

**Reply by lawdank** [reply]
Thank you @ndobb & @artgoldberg

**Reply by lawdank** [reply]
Hi @ndobb  - I executed npm command on our production server with the `--production` flag & could still see the same result :
<img width="627" alt="image" src="https://user-images.githubusercontent.com/26530107/189954671-543ca62e-3159-4987-b6e1-14303e53036a.png">

**Reply by ndobb** [reply]
Right, nothing surprising there @lawdank - unless you're willing to also update the Leaf source code, such warnings will not go away and the app will cease to compile if you force an upgrade. As I mentioned, I have an active branch with those issues resolved, but until we have tested it a bit more it won't be an official release.

**Reply by lawdank** [reply]
Ah... Got it. Thanks @ndobb . We will wait for the official release.
